Best practices/Source Generators.md

```diff
@@ -0,0 +1,147 @@
+## What are Source Generators?
+
+According to official [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) a source generator is a part of the .NET Compiler Platform ("Roslyn") SDK which allows developers to generate source files during compilation and add those files to the compilation object. 
+
+In order for this to be possible, source generators also have the possibility to retrieve a compilation object which contains all the code a developer has written.
+
+The image below shows the process which happens during compilation if a developer uses source generators.
+
+| ![sourceGenerator](/resources/source-generator-visualization.png) |
+|:--:|
+| Figure 1 - Visualization of compilation that includes a Source Generator |
+
+There are multiple advantages of using source generators which include better performance because the source generator is analyzing code during the compilation. An alternative to this approach is runtime reflection which is used during startup of an application and therefore can prolong the startup time. Another advantage is posibility to catch errors during compile time instead of catching them during runtime.
+
+## Source Generator project setup
```

Comment:
I would omit the "Source Generator" part of this title since this whole section is talking about Source generators